### ISSUE TITLE
[Epic] Book Buddy: On-device, spoiler-free reading companion

### ISSUE THREAD

**Issue by srijan-singh** [srijan-singh/book-budy#1]
## Summary
 
Book Buddy is an on-device reading companion for fiction. As the user reads a book chapter by chapter, they can ask questions, request clarification, or casually chat about the story — without ever being shown content from chapters they haven't reached yet. Everything runs locally on constrained hardware (target: ≤700MB RAM for the model layer), with no server round-trip required for core functionality.
 
This issue captures the architecture decisions made so far and breaks the build into iterative, independently shippable subtasks.
 
---
 
## Core product decisions
 
- **Audience**: fiction readers (majority use case) — spoiler-safety is a first-class requirement, not an add-on.
- **Persona**: a warm, encouraging "buddy" tone — not a generic Q&A tool.
- **Modes of interaction**:
  1. Grounded Q&A ("what does this line mean?", "who is this character?")
  2. Casual chat / FAQ (light banter, doesn't need retrieval)
  3. "Catch me up" — summarize the previous chapter on request
---
 
## Architecture overview
 
```
Flutter UI (Dart)
     │  dart:ffi
     ▼
C core (llama.cpp + retrieval/spoiler logic)
     ├─ Generation: small GGUF model (e.g. Qwen2.5-0.5B-Instruct, Q4_K_M)
     ├─ Embedding: all-MiniLM-L6-v2 GGUF (~20MB, 384-dim)
     └─ Local storage: SQLite (chunks, vectors, reading progress, chat history)
```
 
**Key architectural decisions:**
 
- **Frontend**: Flutter (cross-platform: Android, iOS, desktop).
- **Core logic**: C, using llama.cpp directly (already C/C++, no reinvention needed).
- **Bridge**: `dart:ffi` direct binding — not a sidecar/socket process. Near-native call latency, no serialization overhead. `llama_cpp_dart` package provides low-level FFI bindings + a "Managed Isolate" mode built specifically to avoid blocking the UI thread during generation.
- **Threading**: inference runs in a background isolate; tokens stream back to the UI isolate via `Stream`/`SendPort` for a token-by-token "typing" effect without janking the UI.
- **Storage**: SQLite — chunk text + embedding vectors (as BLOBs) + chapter metadata + reading progress + chat history, all in one file per device. Brute-force cosine similarity over a few thousand rows is fast enough at this scale; no dedicated vector DB needed.
- **Spoiler boundary enforcement**: enforced in the **C retrieval layer**, not the Flutter UI. The similarity search filters `chapter_id <= current_progress` before ranking — this is a correctness/security boundary, not just a UI convenience.
---
 
## Why RAG, not fine-tuning, for book content
 
- Fine-tuning a tiny model (270M–500M params) on book text teaches *style/associations*, not verbatim recall — it compresses the book into a lossy statistical impression, risking confident-but-wrong answers (the opposite of "precise").
- Correct approach: keep facts in **context**, not in weights. Retrieve only the relevant paragraphs at query time and let the model synthesize from visible text, constrained by a strict "answer only from provided context" system prompt.
- Fine-tuning (LoRA/SFT) is still useful — but only for **behavior/persona** (buddy tone, "say I don't know if it's not in the text", citation habits), never for injecting book facts.
## RAG pipeline design
 
- **Chunking**: paragraph-based.
  - Min chunk size ~50 tokens → merge upward with next paragraph if smaller.
  - Max chunk size ~350–400 tokens → split at sentence boundaries if larger (embedding models truncate silently past ~512 tokens).
  - Prepend a short anchor (chapter title + prior paragraph's last sentence) to fix ambiguous pronouns — no full token-overlap needed.
  - Metadata per chunk: `chapter_id`, `paragraph_index`, raw text.
- **Retrieval**: top-k = 3–5 chunks, prefer similarity-threshold + max-k cap over a fixed k (avoid diluting a small model's attention with weak matches).
- **Timing**: embed chunks **once, at ingestion time** (when a book is processed), not on the fly per question. Only the user's live question gets embedded at query time.
- **Ingestion location**: do PDF → markdown → chapter → chunk → embed conversion **offline/once per book** (not on-device from a raw PDF at runtime) — ship each book as a pre-processed package. Device only ever does: query embedding + similarity search + generation.
## Model shortlist (fits ≤700MB RAM budget)
 
| Model | Params | Best quant | File size |
|---|---|---|---|
| SmolLM2-135M(-Instruct) | 135M | Q5_K_M / Q8_0 | ~110–145MB |
| SmolLM2-360M(-Instruct) | 360M | Q4_0 / Q3_K_S | ~220–230MB |
| Gemma-3-270M(-it) | 270M | Q4_K_M | ~150–240MB (32K context per Google's model card) |
| Qwen2.5-0.5B-Instruct | 0.5B | Q3_K_M / Q4_K_M | ~430–490MB (32K context) — **current default pick** |
| Embedding: all-MiniLM-L6-v2 | 22M | Q4–Q5 GGUF | ~20MB |
 
Avoid 1B+ models at Q4 and above — weights alone approach/exceed the 700MB budget before accounting for KV cache and app overhead.
 
## Personalization approach (not classic RL)
 
- What's needed is **online personalization**, not full RLHF/PPO (reward model + rollouts is overkill for adapting a small model to one user).
- Recommended: **LoRA adapters + SFT-on-accepted-output or lightweight DPO**, not full-weight updates or PPO.
  - LoRA keeps the base model frozen, trains a small low-rank delta — cheap, reversible, resistant to catastrophic forgetting.
  - Watch for catastrophic forgetting on tiny models; mitigate with low learning rate + small rehearsal set of generic examples.
- Applies to **persona/behavior tuning** (buddy tone, format habits) — not to book content, which always comes fresh from RAG.
## Distribution considerations
 
- **Model delivery**: ~500MB model won't comfortably fit app-store binary size limits — download on first launch (Wi-Fi-gated) rather than bundling.
- **Book packages**: pre-processed (markdown + chunks + embeddings already computed) rather than shipping raw PDFs for on-device parsing.
---
 
### Phase 1 — Core inference plumbing
- [ ] Stand up llama.cpp as a static/shared C library target for Android + iOS + desktop
- [ ] Integrate `llama_cpp_dart` (or hand-rolled FFI bindings) into a bare Flutter project
- [ ] Get a single GGUF model (start with Qwen2.5-0.5B-Instruct Q4_K_M) loading and generating via FFI
- [ ] Implement background-isolate streaming: tokens flow from C → Dart `Stream` → chat UI, without blocking the main isolate
- [ ] Basic chat UI in Flutter (no retrieval yet — just raw model chat) to validate the plumbing end-to-end
### Phase 2 — Book ingestion pipeline (offline/one-time per book)
- [ ] PDF → markdown conversion (external tool, not on-device)
- [ ] Chapter splitting logic
- [ ] Paragraph-based chunker with min/max merge-split rules
- [ ] Embedding generation for all chunks (all-MiniLM-L6-v2 via llama.cpp embedding mode)
- [ ] Define book package format (markdown + chunks + vectors + chapter metadata) for shipping to the app
### Phase 3 — On-device retrieval
- [ ] SQLite schema: chunks table (text, vector BLOB, chapter_id, paragraph_index), reading_progress table, chat_history table
- [ ] Load a book package into SQLite on first open
- [ ] Implement cosine similarity search in C over stored vectors
- [ ] Enforce `chapter_id <= current_progress` filter at the retrieval layer (spoiler boundary)
- [ ] Query embedding at ask-time → retrieval → top-k chunk selection with similarity threshold
### Phase 4 — Grounded Q&A
- [ ] System prompt template: "answer only using provided text, cite when possible, say so if not present"
- [ ] Wire retrieved chunks into the generation prompt
- [ ] Low-temperature decoding for factual questions
- [ ] Manual QA pass: ask questions with known answers/absences, verify no hallucination past chapter boundary
### Phase 5 — Buddy persona & casual chat
- [ ] Collect/label a small set of buddy-toned example responses
- [ ] LoRA fine-tune (or SFT) on persona examples, applied via llama.cpp's LoRA loading
- [ ] Lightweight router: decide per-message whether to hit RAG or answer from immediate conversation context only (avoid running retrieval on casual banter)
- [ ] "Catch me up" mode: summarize previous chapter's chunks on request
### Phase 6 — Reading progress & session state
- [ ] Persist current chapter/progress per book
- [ ] UI for marking progress (manual or chapter-completion-based)
- [ ] Per-chapter/session conversation memory so the buddy doesn't re-explain things already covered
### Phase 7 — Distribution & packaging
- [ ] First-launch model download flow (Wi-Fi gated, progress UI)
- [ ] Book package download/import flow
- [ ] RAM/perf testing on actual low-end target hardware, not just dev machines
---
 
## Open questions / follow-ups
- Exact chunk size tuning (350–400 token ceiling) may need per-book adjustment based on prose density — worth testing against 2–3 real books before locking defaults.
- Router heuristic for "chat vs. retrieval" (Phase 5) needs a concrete implementation decision — confidence threshold from the model itself, or a separate lightweight classifier?
- Confirm exact context window of whichever SmolLM2 variant is used as a fallback model (commonly ~8K — verify before relying on it for longer chapters).

**Comment by srijan-singh** [comment]
UX: [Current UX Design](https://pro-2684.github.io/GitHub-Preview/srijan-singh/book-budy/refs/heads/main/book-buddy-ux.html)

**Comment by srijan-singh** [comment]
PDF → markdown conversion: https://github.com/SubhamTyagi/android-ocr
